### PR TITLE
Add generic functions to encode/decode around a tuple

### PR DESCRIPTION
**Pull request by hansl**
This is follow-up and update to PR #38. It is backward-compatible, but
we may want to deprecate the macros as they are harder to use and non
type safe in some edge cases.

Tests were modified to ensure that the macros and new decode/encode
functions are fully compatible.

Please note that we don't verify that the decoder decoded all values, as decoding should be covariant on arguments.

### CODE REVIEW

File: rust/candid/src/de.rs [hunk 1]
```diff
@@ -843,3 +843,184 @@ impl<'de, 'a> de::VariantAccess<'de> for Compound<'a, 'de> {
...
+/// assert_eq!(golden1, value1);
+/// assert_eq!(golden2, value2);
+/// ```
+pub fn decode_args<'a, Tuple>(bytes: &'a [u8]) -> Result<Tuple>
+where
+    Tuple: ArgumentDecoder<'a>,
+{
+    let mut de = IDLDeserialize::new(bytes)?;
+    let res = ArgumentDecoder::decode(&mut de)?;
+    Ok(res)
```
Comment (chenyan-dfinity) [inline rust/candid/src/de.rs:1004]:
```suggestion
    de.done()?;
    Ok(res)
```
`done()` ensures the remaining bytes are proper Candid encoding, a good safe guard.

File: rust/candid/src/de.rs [hunk 2]
```diff
@@ -843,3 +843,184 @@ impl<'de, 'a> de::VariantAccess<'de> for Compound<'a, 'de> {
...
+/// ```
+/// # use candid::Encode;
+/// # use candid::de::decode_one;
+/// let golden1 = 123u64;
+/// let bytes = Encode!(&golden1).unwrap();
+/// let value1: u64 = decode_one(&bytes).unwrap();
+///
+/// assert_eq!(golden1, value1);
+/// ```
+pub fn decode_one<'a, T>(bytes: &'a [u8]) -> Result<T>
```
Comment (chenyan-dfinity) [inline rust/candid/src/de.rs:1020]: Non-blocking, but I wonder if we can implement `ArgumentDecoder` for any `Deserialize`, so that we can avoid special case for one argument.

File: rust/candid/tests/serde.rs [hunk 3]
```diff
@@ -492,7 +492,11 @@ where
 }
 
 fn encode<T: CandidType>(value: &T) -> Vec<u8> {
```
Comment (chenyan-dfinity) [inline rust/candid/tests/serde.rs:494]: Update the `decode` test as well.

**Review by chenyan-dfinity — COMMENTED** [review]
LGTM. The macro is never intended to be for generic use. It's like `vec!`, convenient to use for fixed interfaces like the management canister. The generic case should always use the builder pattern. Fair enough to provide a convenience wrapper for sequence of arguments.

File: rust/candid/src/de.rs (hunk 2 above)
Comment (hansl) [reply rust/candid/src/de.rs:1020]: Once specialization makes it past unstable, we could use it for exactly that. See https://github.com/rust-lang/rust/issues/31844

Unfortunately the type system is currently limited and trying to do a generic `impl<T> ArgumentDecoder for T` will conflict with the existing implementations that include tuples.

**Comment by hansl** [comment]
@chenyan-dfinity Addressed all your comments. Thanks for the review!

**Review by chenyan-dfinity — APPROVED** [review]
Thanks!
